### ISSUE TITLE
Macro as an exression on non-GCC compiler (expected an expression)

### ISSUE THREAD

**Issue by Si-i3ima** [m3101/The-Powder-Toy#1]
If i understand correctly, the problem is that the macro written for GCC is not correctly recognized by a compiler like MSVC
![image](https://user-images.githubusercontent.com/17309019/89129790-f1b75500-d508-11ea-94f2-6fe63ed4c09e.png)

One of the solutions (at the moment) is using WSL. As you can see, it kinda works if WSL disto have all needed packages installed
![image](https://user-images.githubusercontent.com/17309019/89130514-3ee9f580-d50e-11ea-88fc-2960188cb62a.png)
![image](https://user-images.githubusercontent.com/17309019/89130737-08ad7580-d510-11ea-8cd4-a74e0a972ad4.png)

But when it comes to compiling using only Windows.... Well first of if you using windows you need to install SCons for Windows and add it to PATH, instructions and files are [_HERE_](https://sourceforge.net/projects/scons/files/scons/4.0.1/)
But I met this. Looking for solution, need help
![image](https://user-images.githubusercontent.com/17309019/89131179-7effa700-d513-11ea-8853-43e04825171d.png)

**Comment by m3101** [comment]
Thanks for opening an issue for this problem!
WSL (Windows Subsystem for Linux) is a very elegant solution.
The problem you ran into is that your compiler didn't find one of the libraries the project depends on.
I don't have much experience with Windows development, but I believe [MinGW](http://www.mingw.org/) can provide a reliable GCC build for windows, and also other packages have been built for it (such as SDL2, for graphics, which is where your build failed).

[Here](https://w3.cs.jmu.edu/bernstdh/Web/common/help/cpp_mingw-sdl-setup.php)'s a good guide on installing SDL2 on MinGW.

The downside is that one will have to manually install all the dependencies (which could be installed with a single command on Linux or WSL).

I believe building on linux and testing the `.exe` builds on another Windows system (I believe I still have a HD with Windows 7 somewhere. Windows is very probably backwards compatible that far.) is what I will personally do with the mod. You could try doing the same on your setup (building on WSL and running it on Windows). As for a pure-windows setup, I don't really know of any solution besides MinGW and a lot of patience and library-hunting.

**Comment by Si-i3ima** [comment]
>but I believe [MinGW](http://www.mingw.org/) can provide a reliable GCC build for windows

I found all requires libs and includes on the official wiki of TPT - 
[Required Libraries.zip](https://github.com/m3101/The-Powder-Toy/files/5012951/Required.Libraries.zip)
And extracted it into folders inside mingw directory and compilation started, until some [point](https://del.dog/ylonunitel)

**Comment by m3101** [comment]
Weirdly, the first error is completely outside the modifications I made.
I suspect it's not caused by the mod, but it's still worth investigating. I'll be reading more into that as soon as possible.

The second one failed to find <dirent.h>, a POSIX header. I believe Microsoft didn't provide POSIX compliance either.
There are some implementations of this header around for windows (see some [here](https://stackoverflow.com/questions/5530933/dirent-h-in-visual-studio-2010-or-2008)), but that is kind of a workaround.
Another solution might be [CygWin](https://www.cygwin.com/), a POSIX-compliant environment for Windows. This seems to be the most elegant/correct choice for this situation.

Please let me know if you make any progress on this issue.

**Comment by Si-i3ima** [comment]
Dirent is already in includes, but i still getting same checking error
![image](https://user-images.githubusercontent.com/17309019/89133160-7c587e00-d522-11ea-93c1-5dadad7be74b.png)

Later i'll Cygwin

**Comment by Si-i3ima** [comment]
Hey! Guess what
![image](https://user-images.githubusercontent.com/17309019/89133891-b4fb5600-d528-11ea-88ab-9be65119fff3.png)

**Comment by Si-i3ima** [comment]
Success, but despite the fact that I used --static executable requires one file
![image](https://user-images.githubusercontent.com/17309019/89134382-6b613a00-d52d-11ea-8614-0c6d558097a6.png)

**Comment by Si-i3ima** [comment]
Success, but despite the fact that I used --static executable requires one file - libgcc_s_sjij-1.dll

**Comment by Si-i3ima** [comment]
I fixed it, `env.Append(LINKFLAGS=["-static-libgcc"])` added to line 416 in SConscript, here's exe - 
[Powder.zip](https://github.com/m3101/The-Powder-Toy/files/5016070/Powder.zip)

**Comment by LBPHacker** [comment]
You two should just implement [these](https://github.com/m3101/The-Powder-Toy/commit/f9c760231c4ea3cfd84e5ad59e90583f1f482556#diff-a3cb0a9c2d7f5588ec22bbb7385d06dfR102-R141) as C++ functions and let the compiler optimise them to its liking. Don't try to be smarter than the compiler, it gets you into situations like this.

On windows, `scons` or at worst `scons --msvc` should work if you have msvc installed.

**Comment by m3101** [comment]
@LBPHacker 
Certainly. I'll do that. I just used macros because of a (bad) habit (lambdas and similar things don't exist in C, but C++ seems to have them) and the fact I didn't want to add new .cpp files to the compiling process (I don't really know how the build process currently works, and I was too lazy to study it :/ ).
Anyhow, it's definitely best to just implement them as functions. It makes no sense to avoid that on this context.
Thanks for the suggestions and time.

**Comment by LBPHacker** [comment]
You're very welcome. The current SCons setup just globs every `*.cpp` in `src/` and some other places, so you just add your files wherever and it'll pick it up. (This is its own kind of evil of course, and the next build system config I've emailed you about will do it differently. I.e. don't try this at home.)

**Comment by LBPHacker** [comment]
Oh btw not even the compiler will optimise your functions too well if you put them in their own translation unit (.cpp file) because the only time it'd see the "big picture" would be link time (there's link-time optimisation, but it's not enabled by default). What you could do instead is have them defined `inline` in header files. (`inline` in C++ just allows exceptions to the ODR to be made, it has nothing to do with inlining other than making it more likely to happen when using the vast majority of compilers without LTO.)
